pkg/validation/events.go

```diff
@@ -0,0 +1,136 @@
+		fallthrough
+	case len(events.PostStart) > 0:
+		if postStartErr := isEventValid(events.PostStart, postStart, commandMap); postStartErr != nil {
+			eventErrors += fmt.Sprintf("\n%s", postStartErr.Error())
+		}
+		fallthrough
+	case len(events.PreStop) > 0:
+		if preStopErr := isEventValid(events.PreStop, preStop, commandMap); preStopErr != nil {
+			eventErrors += fmt.Sprintf("\n%s", preStopErr.Error())
+		}
+		fallthrough
+	case len(events.PostStop) > 0:
+		if postStopErr := isEventValid(events.PostStop, postStop, commandMap); postStopErr != nil {
+			eventErrors += fmt.Sprintf("\n%s", postStopErr.Error())
+		}
+	}
+
+	// if there is any validation error, return it
+	if len(eventErrors) > 0 {
+		return fmt.Errorf("devfile events validation error: %s", eventErrors)
+	}
+
+	return nil
+}
+
+// isEventValid checks if events belonging to a specific event type are valid ie;
+// 1. event should map to a valid devfile command
+// 2. preStart and postStop events should either map to an apply command or a composite command with apply commands
+// 3. postStart and preStop events should either map to an exec command or a composite command with exec commands
+func isEventValid(eventNames []string, eventType string, commandMap map[string]v1alpha2.Command) error {
+	var invalidCommand, applyEvents, execEvents []string
+
+	for _, eventName := range eventNames {
+		if command, ok := commandMap[strings.ToLower(eventName)]; !ok { // check if event is in the list of devfile commands
+			invalidCommand = append(invalidCommand, eventName)
+		} else {
+			switch eventType {
+			case preStart:
+				fallthrough
+			case postStop:
+				// check if the event is either an apply command or a composite of apply commands
+				if command.Apply == nil && command.Composite == nil {
+					applyEvents = append(applyEvents, eventName)
+				} else if command.Composite != nil {
+					applyEvents = append(applyEvents, validateCompositeEvents(*command.Composite, eventName, eventType, commandMap)...)
+				}
+			case postStart:
+				fallthrough
+			case preStop:
+				// check if the event is either an exec command or a composite of exec commands
+				if command.Exec == nil && command.Composite == nil {
+					execEvents = append(execEvents, eventName)
+				} else if command.Composite != nil {
+					execEvents = append(execEvents, validateCompositeEvents(*command.Composite, eventName, eventType, commandMap)...)
+				}
+			}
+		}
+	}
+
+	var eventErrors string
+	var err error
+
+	if len(invalidCommand) > 0 {
+		eventErrors = fmt.Sprintf("\n%s does not map to a valid devfile command", strings.Join(invalidCommand, ", "))
+	}
+
+	if len(applyEvents) > 0 {
+		eventErrors += fmt.Sprintf("\n%s should either map to an apply command or a composite command with apply commands", strings.Join(applyEvents, ", "))
+	}
+
+	if len(execEvents) > 0 {
+		eventErrors += fmt.Sprintf("\n%s should either map to an exec command or a composite command with exec commands", strings.Join(execEvents, ", "))
+	}
```

Comment:
`applyEvents` and `execEvents` is unclear in this case, consider renaming to indicate that these are lists of apply/exec events added in error